libmogwai-schedule/tests/clock-dummy.c

```diff
@@ -0,0 +1,412 @@
+ * whose trigger times are earlier than or equal to @now being triggered. Any
+ * new alarms added from alarm callbacks are also invoked if their trigger times
+ * meet the same criterion.
+ *
+ * The time zone for @now is not used to change the time zone for the clock;
+ * @now is converted to the current time zone in use by the clock. To change the
+ * clock’s time zone, call mws_clock_dummy_set_time_zone().
+ *
+ * Since: 0.1.0
+ */
+void
+mws_clock_dummy_set_time (MwsClockDummy *self,
+                          GDateTime     *now)
+{
+  g_return_if_fail (MWS_IS_CLOCK_DUMMY (self));
+  g_return_if_fail (now != NULL);
+
+  g_autofree gchar *now_str = g_date_time_format (now, "%FT%T%:::z");
+  g_debug ("%s: Setting time to %s; %u alarms to check",
+           G_STRFUNC, now_str, self->alarms->len);
+
+  /* Trigger all the alarms before the new @now (inclusive).
+   * Since other methods on #MwsClockDummy may be called by the user function
+   * in alarm_data_invoke(), which may modify the @self->alarms array, we need
+   * to be careful to re-read the array on each loop iteration, and to remove
+   * invoked alarms individually. */
+  while (self->alarms->len > 0)
+    {
+      g_autoptr(AlarmData) owned_data = NULL;
+      const AlarmData *data = g_ptr_array_index (self->alarms, 0);
+
+      g_autofree gchar *alarm_time_str = g_date_time_format (data->alarm_time, "%FT%T%:::z");
+      g_debug ("%s: Comparing alarm %s to now %s",
+               G_STRFUNC, alarm_time_str, now_str);
+
+      if (g_date_time_compare (data->alarm_time, now) > 0)
+        break;
+
+      /* Steal the alarm from the array.
+       * FIXME: Use g_ptr_array_steal_index() when we can depend on a suitable GLib version. */
+      owned_data = g_steal_pointer (&self->alarms->pdata[0]);
+      g_ptr_array_remove_index (self->alarms, 0);
+
+      /* Set the current time to what the alarm expects. */
+      g_date_time_unref (self->now);
+      self->now = g_date_time_to_timezone (owned_data->alarm_time, self->tz);
+
+      alarm_data_invoke (owned_data);
+    }
+
+  /* Convert the final time to our local timezone. */
+  g_date_time_unref (self->now);
+  self->now = g_date_time_to_timezone (now, self->tz);
+}
+
+/**
+ * mws_clock_dummy_set_time_zone:
+ * @self: a #MwsClockDummy
+ * @tz: new time zone for the clock
+ *
+ * Set the clock’s time zone to @tz, and convert its current ‘now’ time to be
+ * the same instant as before, but in @tz.
+ *
+ * If the time zone has changed, this results in the #MwsClock::offset-changed
+ * signal being emitted.
+ *
+ * Since: 0.1.0
+ */
+void
+mws_clock_dummy_set_time_zone (MwsClockDummy *self,
+                               GTimeZone     *tz)
+{
+  g_return_if_fail (MWS_IS_CLOCK_DUMMY (self));
+  g_return_if_fail (tz != NULL);
+
+  /* FIXME: Do we need a g_time_zone_equal() function to compare these? This
+   * is a terrible way of comparing them. */
+  if (tz == self->tz)
+    return;
```

Comment:
Comparing by identifiers is more complex than it’s worth, as you have to provide an instant to look up the identifiers in (because, for example, in the UK there is a different identifier for summer time vs winter time). I think identifiers are too fragile.

Offsets could result in false positives, as two timezones could share an offset if one of them is in summer time and the other in winter time. I think that is also too fragile.

I’ll leave it as a pointer comparison for now, and eventually hope to get new API into GLib to allow proper comparisons.